### PR TITLE
DEPS: update setuptools min version

### PR DESCRIPTION
**Pull request by fangchenli**
According to https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html, the min version of setuptools that support `entry_points` in `setup.cfg` is  `51.0.0`.

### CODE REVIEW

**Comment by jreback** [comment]
cc @simonjayhawkins

**Comment by simonjayhawkins** [comment]
are we not testing with the minimum setuptools?

**Comment by fangchenli** [comment]
> are we not testing with the minimum setuptools?

No, we are not.

**Comment by jreback** [comment]
can you update the min pin in environment.yaml (and other places if needed)?

**Comment by simonjayhawkins** [comment]
test failure unrelated

```
E           AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed <ssl.SSLSocket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.1.1.54', 48258), raddr=('52.217.136.144', 443)>"), '/home/runner/work/pandas/pandas/pandas/io/sas/sas7bdat.py', 381)]
```

**Comment by simonjayhawkins** [comment]
> can you update the min pin in environment.yaml (and other places if needed)?

@fangchenli can you address this

**Comment by simonjayhawkins** [comment]
> also this should be added to the wheel recipe @simonjayhawkins (practically this already works, but should add)

pandas-feedstock or MacPython or both?

**Comment by fangchenli** [comment]
> can you update the min pin in environment.yaml (and other places if needed)?

We don't list setuptools in `environment.yaml`.

**Comment by simonjayhawkins** [comment]
labelled as blocker as we won't be changing min versions of dependencies in patch releases?

**Comment by jreback** [comment]
yeah i think we should just add this to `environment.yaml`. it is technically only needed if we use matplotlib so its not a required dep (so only add it where we have matplotlib, which is most places).

**Comment by simonjayhawkins** [comment]
conda-incubator/setup-miniconda@v2 does not seem to like the setuptools pin, although I _think_ there have been conda-incubator/setup-miniconda@v2 timeouts on other PRs

**Comment by simonjayhawkins** [comment]
The version in pandas-dev on CI / Checks for master (https://github.com/pandas-dev/pandas/runs/2950514145) is

```
2021-06-30T09:08:01.2373051Z setuptools                49.6.0           py38h578d9bd_3    conda-forge
```

and creating dev env locally

my backport env (1.3.x)
```
(pandas-backport) simon@ProDesk:~/pandas (backport-of-pr-42318-on-1.3.x)$ conda list setuptools
# packages in environment at /home/simon/miniconda3/envs/pandas-backport:
#
# Name                    Version                   Build  Channel
setuptools                49.6.0           py38h578d9bd_3    conda-forge
```

my master env
```
(pandas-dev) simon@P340:~/pandas (master)$ conda list setuptools
# packages in environment at /home/simon/miniconda3/envs/pandas-dev:
#
# Name                    Version                   Build  Channel
setuptools                49.6.0           py38h578d9bd_3    conda-forge
```

> According to https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html, the min version of setuptools that support `entry_points` in `setup.cfg` is `51.0.0`.

hmm? our tests pass locally?

**Comment by simonjayhawkins** [comment]
@jreback ok to release without this PR, or delay release to allow time for others to look at this? @pandas-dev/pandas-core

**Comment by jreback** [comment]
should
include this

**Comment by simonjayhawkins** [comment]
> should
> include this

it appears can't a dev env with conda with this pin

**Comment by bashtage** [comment]
Could leave it out of the yml file and then do a pip install -r requirements-dev.txt (assuming it is in requirements-dev.txt). This will no op everything that is installed, but upgrade setuptools if below.

**Comment by simonjayhawkins** [comment]
> > should
> > include this
> 
> it appears can't create a dev env with conda with this pin

can create a dev env locally from scratch using this pin. so maybe something related to  conda-incubator/setup-miniconda@v2

creating a dev env from scratch without this pin, the only diff is

```
< setuptools=49.6.0=py38h578d9bd_3
---
> setuptools=52.0.0=py38h06a4308_0
```

so it's a mystery to me how conda chooses the version.

**Comment by jreback** [comment]
hmm yeah this should just work.

File: pyproject.toml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # Minimum requirements for the build system to execute.
 # See https://github.com/scipy/scipy/pull/12940 for the AIX issue.
 requires = [
-    "setuptools>=38.6.0",
+    "setuptools>=51.0.0",
```
Comment (jreback) [inline pyproject.toml:5]: do we also need to change environment.yaml (or was that a difffernt issue)?

**Comment by simonjayhawkins** [comment]
@fangchenli can you summarize the issue this PR is trying to solve

File: doc/source/whatsnew/v1.3.0.rst [hunk 2]
```diff
@@ -645,7 +645,7 @@ If installed, we now require:
 +-----------------+-----------------+----------+---------+
 | mypy (dev)      | 0.812           |          |    X    |
 +-----------------+-----------------+----------+---------+
-| setuptools      | 38.6.0          |          |    X    |
+| setuptools      | 51.0.0          |          |    X    |
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.3.0.rst:648]: is this for users building pandas from source or users using a pre-built pandas with matplotlib?

File: doc/source/getting_started/install.rst [hunk 3]
```diff
@@ -262,7 +262,7 @@ Visualization
 ========================= ================== =============================================================
 Dependency                Minimum Version    Notes
 ========================= ================== =============================================================
-setuptools                38.6.0             Utils for entry points of plotting backend
+setuptools                51.0.0             Utils for entry points of plotting backend
```
Comment (simonjayhawkins) [inline doc/source/getting_started/install.rst:265]: the notes suggest the min version we require  is for using the plotting backend.

**Comment by simonjayhawkins** [comment]
> > are we not testing with the minimum setuptools?
> 
> No, we are not.

@fangchenli should we test with the min version?

> yeah i think we should just add this to `environment.yaml`. it is technically only needed if we use matplotlib so its not a required dep (so only add it where we have matplotlib, which is most places).

and have a min pin in the envs with matplotlib?

**Comment by jorisvandenbossche** [comment]
> According to https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html, the min version of setuptools that support entry_points in setup.cfg is 51.0.0.

I think we need to distinguish the install-time and run-time dependency here. For installing pandas with setting up the entrypoint, we need a recent setuptools (since we are using `setup.cfg` now), but for *running* pandas and using setuptools to find the entry point for a given plotting backend, we use the `setuptools.pkg_resources` submodule. I think this module is already existing much longer in setuptools, and I don't think we need a minimum version for setuptools as a run-time dependency.

For setuptools as a build/install-time dependency, we only need to update the pyproject.toml, I think:

File: doc/source/development/contributing_environment.rst [hunk 4]
```diff
@@ -189,9 +189,8 @@ Creating a Python environment (pip)
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 If you aren't using conda for your development environment, follow these instructions.
-You'll need to have at least the :ref:`minimum Python version <install.version>` that pandas supports. If your Python version
-is 3.8.0 (or later), you might need to update your ``setuptools`` to version 42.0.0 (or later)
-in your development environment before installing the build dependencies::
+You'll need to have at least the :ref:`minimum Python version <install.version>` that pandas supports.
+In your development environment before installing the build dependencies::
```
Comment (jorisvandenbossche) [inline doc/source/development/contributing_environment.rst:193]: This sentence is no longer a complete sentence.

File: doc/source/getting_started/install.rst (hunk 3 above)
Comment (jorisvandenbossche) [reply doc/source/getting_started/install.rst:265]: Yes, this doesn't need to be changed.

**Comment by bashtage** [comment]
>  I think this module is already existing much longer in setuptools, and I don't think we need a minimum version for setuptools as a run-time dependency.

If setuptools isn't listed as a run-time dependency at all, it is possible that an install might not have it.  It looks like pkg_resources is imported in an unprotected manner here

https://github.com/pandas-dev/pandas/blob/059c8bac51e47d6eaaa3e36d6a293a22312925e6/pandas/tests/plotting/test_backend.py#L4

which I think means test discovery will fail.  

Should this import be moved inside the function that uses it?

**Comment by bashtage** [comment]
> Why I am not getting any failures? are we missing tests?

AFAICT the only way to get a failure (which will be an error in test discovery) is to run the tests in an environment without setuptools.  This is possible, but in most cases would need a deliberate attempt to create (i.e., install pandas, remove setuptools, run tests).

**Comment by fangchenli** [comment]
> For setuptools as a build/install-time dependency, we only need to update the pyproject.toml, I think.

I agree. Also, we don't know what's the min version of setuptools that supports the function we used in matplotlib backend. The previous min version `38.6.0` was also determined by an attribute we used in setup.cfg.

File: doc/source/development/contributing_environment.rst [hunk 5]
```diff
@@ -189,11 +189,8 @@ Creating a Python environment (pip)
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 If you aren't using conda for your development environment, follow these instructions.
-You'll need to have at least the :ref:`minimum Python version <install.version>` that pandas supports. If your Python version
-is 3.8.0 (or later), you might need to update your ``setuptools`` to version 42.0.0 (or later)
-in your development environment before installing the build dependencies::
-
-      pip install --upgrade setuptools
+You'll need to have at least the :ref:`minimum Python version <install.version>` that pandas supports.
+You also need to have ``setuotools`` 51.0.0 or later to build pandas.
```
Comment (simonjayhawkins) [inline doc/source/development/contributing_environment.rst:193]:
```suggestion
You also need to have ``setuptools`` 51.0.0 or later to build pandas.
```

**Comment by bashtage** [comment]
> I agree. Also, we don't know what's the min version of setuptools that supports the function we used in matplotlib backend. The previous min version `38.6.0` was also determined by an attribute we used in setup.cfg.

Is there some reason not to lazily import it in the test file where it appears? If it is lazily imported when only will cause a problem is someone (a) has matplotlib and (b) does not have setuptools.

**Comment by simonjayhawkins** [comment]
> labelled as blocker as we won't be changing min versions of dependencies in patch releases?

if we are no longer changing the min version for users, this is no longer a blocker? will release tomorrow even if this PR is not ready?

**Comment by jreback** [comment]
yep let's just release
this is effectively docs at this point

**Comment by simonjayhawkins** [comment]
not yet triaged the pypy failures on conda-forge https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=345935&view=results, but did notice setuptools was pre 0.51

File: doc/source/development/contributing_environment.rst [hunk 6]
```diff
@@ -189,11 +189,8 @@ Creating a Python environment (pip)
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 If you aren't using conda for your development environment, follow these instructions.
-You'll need to have at least the :ref:`minimum Python version <install.version>` that pandas supports. If your Python version
-is 3.8.0 (or later), you might need to update your ``setuptools`` to version 42.0.0 (or later)
-in your development environment before installing the build dependencies::
-
-      pip install --upgrade setuptools
```
Comment (jorisvandenbossche) [inline doc/source/development/contributing_environment.rst:196]: Can you include this step in the instructions block below (eg after creating and activating the virtualenv). Or otherwise we can still include it in the requirements-dev.txt?

File: pyproject.toml (hunk 1 above)
Comment (jreback) [reply pyproject.toml:5]: actually can we remove this now?

Comment (jorisvandenbossche) [reply pyproject.toml:5]: The reason for the higher setuptools version was the ability to read certain information from setup.cfg, I think. So that's still something we need.

Comment (jreback) [reply pyproject.toml:5]: good point ok then!

File: doc/source/development/contributing_environment.rst (hunk 6 above)
Comment (simonjayhawkins) [reply doc/source/development/contributing_environment.rst:196]: > Or otherwise we can still include it in the requirements-dev.txt?

the instructions become redundant if included in requirements-dev.txt? +1 on including and removing the additional instructions

**Comment by simonjayhawkins** [comment]
moving to 1.3.2

**Comment by jreback** [comment]
thanks @fangchenli this is essentially a doc update at this point so backporting ok.

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x
